bin/shared.sh

```diff
@@ -37,50 +34,72 @@ function retry_upgrade {
-  # more likely to be confused by them than helped.
-  #
-  # For "flock", the lock is released when the file descriptor goes out of
-  # scope,  i.e. when this function returns.  The lock is released via
-  # a trap when using "shlock".
+# Determines which lock method to use, based on what is available on the system.
+# Returns a non-zero value if the lock was not acquired, zero if acquired.
+function _lock () {
   if hash flock 2>/dev/null; then
-    flock 3 2>/dev/null || true
-  elif hash shlock 2>/dev/null; then
-    FLUTTER_UPGRADE_LOCK="$FLUTTER_ROOT/bin/cache/.upgrade_lock"
-    while ! shlock -f "$FLUTTER_UPGRADE_LOCK" -p $$ ; do sleep .1 ; done
-    trap _rmlock EXIT
+    flock --nonblock --exclusive 7 2>/dev/null
+  else
+    mkdir "$1" 2>/dev/null
   fi
+}
+
+# Waits for an update lock to be acquired.
+#
+# To ensure that we don't simultaneously update Dart in multiple parallel
+# instances, we try to obtain an exclusive lock on this file descriptor (and
+# thus this script's source file) while we are updating Dart and compiling the
+# script. To do this, we try to use the command line program "flock", which is
+# available on many Unix-like platforms, in particular on most Linux
+# distributions. You give it a file descriptor, and it locks the corresponding
+# file, having inherited the file descriptor from the shell.
+#
+# Complicating matters, there are two major scenarios where this will not
+# work.
+#
+# The first is if the platform doesn't have "flock", for example on macOS. There
+# is not a direct equivalent, so on platforms that don't have flock, we fall
+# back to using mkdir as an atomic operation to create a lock directory. If
+# mkdir is able to create the directory, then the lock is acquired. To determine
+# if we have "flock" available, we use the "hash" shell built-in.
+#
+# The second complication is NFS. On NFS, to obtain an exclusive lock you need a
+# file descriptor that is open for writing. Thus, we ignore errors from flock by
+# redirecting all output to /dev/null, since users will typically not care about
+# errors from flock and are more likely to be confused by them than helped.
+#
+# The upgrade_flutter function calling _wait_for_lock is executed in a subshell
+# with a redirect that pipes the source of this script into file descriptor 7.
+# A flock lock is released when this subshell exits and file descriptor 7 is
+# closed. The mkdir lock is released via an exit trap from the subshell that
+# deletes the lock directory.
+function _wait_for_lock () {
+  FLUTTER_UPGRADE_LOCK="$FLUTTER_ROOT/bin/cache/.upgrade_lock"
+  local waiting_message_displayed
+  while ! _lock "$FLUTTER_UPGRADE_LOCK"; do
+    if [[ -z $waiting_message_displayed ]]; then
+      printf "Waiting for another flutter command to release the startup lock...\r";
+      waiting_message_displayed="true"
+    fi
+    sleep .1;
+  done
+  # If the lock file is acquired, make sure that it is removed on exit.
```

Comment:
nit. should we print a newline here? when I issued `dart -h` while another process had a lock, it looked like this:

```
~/git/flutter$ bin/dart -h
A command-line utility for Dart development.se the startup lock...
```

Reply:
The reason this prints with a return is that the Dart code itself will start up and lock, printing that same message.

So, with the newline, it can look like this:
```
~/git/flutter$ bin/dart -h
Waiting for another flutter command to release the startup lock...
Waiting for another flutter command to release the startup lock...
A command-line utility for Dart development.
```

Maybe printing it twice is better than printing it once and sometimes overwriting it, but...

I had an idea: the code now prints
```bash
printf "                                                                   \r"
```
when this lock is released, and then if the Dart code locks, it'll just flicker once, and then there's no confusion.  Does that work?